File: ios/Sources/PoltioSDK/PoltioWidgetCache.swift
import Foundation

/// Cached outcome of a widget resolution request for a target URL.
public enum CachedWidgetResult: Equatable {
    /// Successfully resolved widget metadata.
    case widget(PoltioWidgetResponse)
    /// Explicitly verified that no widget is configured for this URL (404 Not Found).
    case noWidget
}

/// Thread-safe in-memory cache entry with expiration timestamp.
final class PoltioCacheEntry: NSObject {
    let result: CachedWidgetResult
    let expirationDate: Date

    init(result: CachedWidgetResult, ttl: TimeInterval) {
        self.result = result
        expirationDate = Date().addingTimeInterval(ttl)
    }

    var isExpired: Bool {
        Date() >= expirationDate
    }
}

/// Lightweight, in-memory ephemeral cache manager for widget resolutions.
/// Automatically limits memory footprint via `NSCache` and expires entries after the configured TTL.
final class PoltioWidgetCache {
    private let cache = NSCache<NSString, PoltioCacheEntry>()
    private let lock = NSLock()
    private var _defaultTTL: TimeInterval

    /// Default Time-To-Live in seconds for cached widget responses (default is 300 seconds / 5 minutes).
    var defaultTTL: TimeInterval {
        get {
            lock.lock()
            defer { lock.unlock() }
            return _defaultTTL
        }
        set {
            lock.lock()
            defer { lock.unlock() }
            _defaultTTL = newValue
        }
    }

    /// Maximum number of items retained in the cache before older items are evicted (default is 100).
    var countLimit: Int {
        get { cache.countLimit }
        set { cache.countLimit = newValue }
    }

    /// Initializes a new widget cache with specified default TTL and item count limit.
    /// - Parameters:
    ///   - defaultTTL: Cache duration in seconds (defaults to 300.0s / 5 minutes).
    ///   - countLimit: Maximum number of entries (defaults to 100).
    init(defaultTTL: TimeInterval = 300.0, countLimit: Int = 100) {
        _defaultTTL = defaultTTL
        cache.countLimit = countLimit
    }

    /// Retrieves the cached widget result for the specified URL if present and unexpired.
    /// - Parameter url: The target URL string.
    /// - Returns: The cached `CachedWidgetResult`, or `nil` on cache miss / expiration.
    func get(for url: String) -> CachedWidgetResult? {
        let ttl = defaultTTL
        guard ttl > 0 else { return nil }

        let key = url as NSString
        guard let entry = cache.object(forKey: key) else {
            return nil
        }

        if entry.isExpired {
            cache.removeObject(forKey: key)
            PoltioLogger.debug("Cache entry expired for '\(url)'.")
            return nil
        }

        return entry.result
    }

    /// Stores a widget resolution outcome in the cache.
    /// - Parameters:
    ///   - result: The resolution outcome (`.widget` or `.noWidget`).
    ///   - url: The target URL string.
    ///   - ttl: Optional override TTL in seconds (uses `defaultTTL` if nil).
    func set(result: CachedWidgetResult, for url: String, ttl: TimeInterval? = nil) {
        let effectiveTTL = ttl ?? defaultTTL
        guard effectiveTTL > 0 else { return }

        let entry = PoltioCacheEntry(result: result, ttl: effectiveTTL)
        cache.setObject(entry, forKey: url as NSString)
        PoltioLogger.debug("Cached widget resolution for '\(url)' (TTL: \(Int(effectiveTTL))s).")
    }

    /// Removes the cached entry for a specific URL.
    /// - Parameter url: The target URL string.
    func remove(for url: String) {
        cache.removeObject(forKey: url as NSString)
    }

    /// Clears all entries from the in-memory cache.
    func clear() {
        cache.removeAllObjects()
        PoltioLogger.debug("Widget cache cleared.")
    }
}

File: android/poltio-sdk/src/main/java/com/poltio/sdk/PoltioWidgetCache.kt
package com.poltio.sdk

/** Cached outcome of a widget resolution request for a target URL. */
internal sealed class CachedWidgetResult {
    /** Successfully resolved widget metadata. */
    data class Widget(val response: PoltioWidgetResponse) : CachedWidgetResult()

    /** Explicitly verified that no widget is configured for this URL (404 Not Found). */
    object NoWidget : CachedWidgetResult()
}

/**
 * Lightweight, in-memory ephemeral cache manager for widget resolutions. Automatically limits its
 * memory footprint via an access-ordered LRU map and expires entries after the configured TTL.
 */
internal class PoltioWidgetCache(
    defaultTTLSeconds: Double = 300.0,
    countLimit: Int = 100,
) {
    private data class Entry(val result: CachedWidgetResult, val expiresAtMs: Long)

    private val lock = Any()

    /** Default Time-To-Live in seconds for cached widget responses (default is 300s / 5 minutes). */
    @Volatile
    var defaultTTL: Double = defaultTTLSeconds

    /** Maximum number of items retained in the cache before older items are evicted (default is 100). */
    var countLimit: Int = countLimit
        set(value) {
            field = value
            synchronized(lock) { trimToLimit() }
        }

    private val map = object : LinkedHashMap<String, Entry>(16, 0.75f, true) {
        override fun removeEldestEntry(eldest: MutableMap.MutableEntry<String, Entry>): Boolean {
            return size > countLimit
        }
    }

    /** Retrieves the cached widget result for the specified URL if present and unexpired. */
    fun get(url: String): CachedWidgetResult? {
        val ttl = defaultTTL
        if (ttl <= 0) return null

        synchronized(lock) {
            val entry = map[url] ?: return null
            if (System.currentTimeMillis() >= entry.expiresAtMs) {
                map.remove(url)
                PoltioLogger.debug { "Cache entry expired for '$url'." }
                return null
            }
            return entry.result
        }
    }

    /** Stores a widget resolution outcome in the cache. */
    fun set(result: CachedWidgetResult, url: String, ttlSeconds: Double? = null) {
        val effectiveTTL = ttlSeconds ?: defaultTTL
        if (effectiveTTL <= 0) return

        synchronized(lock) {
            map[url] = Entry(result, System.currentTimeMillis() + (effectiveTTL * 1000).toLong())
        }
        PoltioLogger.debug { "Cached widget resolution for '$url' (TTL: ${effectiveTTL.toInt()}s)." }
    }

    /** Removes the cached entry for a specific URL. */
    fun remove(url: String) {
        synchronized(lock) { map.remove(url) }
    }

    /** Clears all entries from the in-memory cache. */
    fun clear() {
        synchronized(lock) { map.clear() }
        PoltioLogger.debug { "Widget cache cleared." }
    }

    private fun trimToLimit() {
        val iterator = map.entries.iterator()
        while (map.size > countLimit && iterator.hasNext()) {
            iterator.next()
            iterator.remove()
        }
    }
}
